lib/filterx/filterx-metrics-labels.c

```diff
@@ -0,0 +1,348 @@
+/*
+ * Copyright (c) 2024 Axoflow
+ * Copyright (c) 2024 Attila Szakacs <attila.szakacs@axoflow.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "filterx-metrics-labels.h"
+#include "filterx-eval.h"
+#include "expr-literal.h"
+#include "expr-literal-generator.h"
+#include "object-string.h"
+#include "object-dict-interface.h"
+#include "metrics/metrics-tls-cache.h"
+#include "stats/stats-cluster-single.h"
+#include "scratch-buffers.h"
+
+#include <string.h>
+
+typedef struct _FilterXMetricsLabel
+{
+  gchar *name;
+  struct
+  {
+    FilterXExpr *expr;
+    gchar *str;
+  } value;
+} FilterXMetricsLabel;
+
+static gchar *
+_format_str_obj(FilterXObject *obj)
+{
+  GString *str = scratch_buffers_alloc();
+  return filterx_object_repr(obj, str) ? str->str : NULL;
+}
+
+static gchar *
+_format_value_expr(FilterXExpr *expr)
+{
+  FilterXObject *obj = filterx_expr_eval_typed(expr);
+  if (!obj)
+    return NULL;
+
+  gchar *result = _format_str_obj(obj);
```

Comment:
Is this okay that we accept everything here?
For example, a deeply nested dict may not be something we want as a label value.
It's probably not a real issue as almost everything is accepted as a label value by Prometheus.